### ISSUE TITLE
Transposition of CSR matrices

### ISSUE THREAD

**Issue by cpuhrsch** [pytorch/pytorch#61059]
## 🚀 Feature
This issue describes and tracks the development of faster transposition kernels for SparseCSR on both CPU and CUDA.

## Motivation

Transposition is a fundamental operation that is often required to speed up matrix multiplication, the backwards pass of matrix multiplication or for bidirectional message passing flow in GNNs. With CSR matrices being applicable to GNNs (e.g. pytoch_sparse) and Transformers (e.g. sparse attention), which are both popular models, we should provide a fast implementation for this operation.

Transpose is a view operation in PyTorch. We currently don't have metadata extensions for CSR to represent transpositions as inexpensive metadata operations such as a CSC tensor or a "transpose" flag. We must implement transposition as a view as it will otherwise yield issues in subsequent autograd operations. (In fact, COO is currently implemented as a copy ([clone() + transpose_](https://github.com/pytorch/pytorch/blob/c50c77b444812d6f28bb7a9071ca0eb643ab6f78/aten/src/ATen/native/TensorShape.cpp#L1697-L1700)), which should be a bug).

EDIT: The main usecase for fast transposition, for now, is the backwards pass of sparse matmul. We have plenty of work to do for sparse matmul and should finish it up first before we continue on with this work. This will then also give us a concrete use site and opportunity to benchmark improvements within transposition.

## Pitch

Implement rudimentary version of [```transpose.int```](https://github.com/pytorch/pytorch/blob/c50c77b444812d6f28bb7a9071ca0eb643ab6f78/aten/src/ATen/native/TensorShape.cpp#L1689-L1713), [```transpose_```](https://github.com/pytorch/pytorch/blob/c50c77b444812d6f28bb7a9071ca0eb643ab6f78/aten/src/ATen/native/TensorShape.cpp#L1665-L1687), [```t```](https://github.com/pytorch/pytorch/blob/c50c77b444812d6f28bb7a9071ca0eb643ab6f78/aten/src/ATen/native/TensorShape.cpp#L1728-L1731) and [```t_```](https://github.com/pytorch/pytorch/blob/c50c77b444812d6f28bb7a9071ca0eb643ab6f78/aten/src/ATen/native/TensorShape.cpp#L1733-L1736) for SparseCsrCPU and SparseCsrCUDA Tensors by modifying the crow_indices and col_indices only and creating a view of the values. Write tests to confirm correctness of forward and autograd. Write benchmarks to establish baseline performance and investigate options for speedup.

EDIT: CSR metadata extensions are mandatory and simply modifying crow and col indicies won't be enough. See discussion below.

## Alternatives

Currently this gap is closed by one of our partner libraries [pytorch_sparse](https://github.com/rusty1s/pytorch_sparse).

## Additional context

```
import torch
crow_indices = torch.tensor([0, 2, 4])
col_indices = torch.tensor([0, 1, 0, 1])
values = torch.tensor([1, 2, 3, 4], dtype=torch.float)
tensor0 = torch.sparse_csr_tensor(crow_indices, col_indices, values)
tensor0.transpose(0, 1)
```

yields

```
Traceback (most recent call last):
  File "/data/home/cpuhrsch/tmp/csr_transpose_bench.py", line 7, in <module>
    tensor0.transpose(0, 1)
NotImplementedError: Could not run 'aten::as_strided' with arguments from the 'SparseCsrCPU' backend
```

Similarly for CUDA we have

```
import torch

crow_indices = torch.tensor([0, 2, 4])
col_indices = torch.tensor([0, 1, 0, 1])
values = torch.tensor([1, 2, 3, 4], dtype=torch.float)
tensor0 = torch.sparse_csr_tensor(crow_indices, col_indices, values, device=torch.device('cuda'))
tensor0.transpose(0, 1)
```

yields

```
Traceback (most recent call last):
  File "/data/home/cpuhrsch/tmp/csr_transpose_bench.py", line 7, in <module>
    tensor0.transpose(0, 1)
NotImplementedError: Could not run 'aten::as_strided' with arguments from the 'SparseCsrCUDA' backend.
```

cc @albanD @IvanYashchuk @pearu @mruberry @bhosmer

**Comment by pearu** [comment]
> Transpose is a view operation in PyTorch.

This statement requires clarification: is it a general requirement or is it just a consequence of using strided layout?

The statement holds for strided tensors where transpose merely recomputes the strides and the result shares the storage of values. For sparse tensors, we don't have strides defined (but see option 4. belwo) and we specify indices explicitly, that is, indices are a considerable part of the tensor data. However, many ways exist to make transpose a view operation for sparse tensor layouts as well:

1. introduce so-called transpose bit to tensor metadata, analogously to recently introduced conjugate bit. Perhaps even unify transpose and conjugate bits frameworks? The advantage of this approach is that it can be used for any layout (CSR, COO, etc). But be aware of risks such as https://github.com/pytorch/pytorch/issues/59945 .
2. for CSR layout, introduce independent CSC layout so that transpose becomes a layout transformation CSR->CSC (and vice-versa) with the same indices/values data. However, this requires that CSR/CSC tensors need to have an owner bit of indices/values data.
3. for CSR layout, extend the layout to represent also CSC layout by introducing a transpose bit to the CSR layout, or just introduce a new CSC layout that shares the implementation with the current CSR implementation (as the corresponding implementations would be ridiculous similar). This approach is conceptually the same as 1. but would have differences in where the transpose bit is stored: in a tensor metadata or in a layout.
4. introduce strides to sparse tensors that function as generalized strides. Generalized strides represent a mapping between tensor indices and the indices of storage tensors. A prototype of the dimensions mapping concept is [available here](https://github.com/pearu/gcs/blob/b54ba0cba9c853b797274ef26b6c42386f2cafa3/gcs/storage.py#L953). As a result, the transpose operation becomes an operation that recomputes strides the same way as for strided tensors. This approach has the advantage that it can be used for any layout as well as it will provide a way to generalize 2-dimensional CSR tensors to N-dimensional tensors via dimensions mapping. The latter may not be required for this issue and one can use the stripped-down version of the generalized strides where the dimensionality is constant. Even then, the approach has an extra value as it would support slicing operations as view operations for sparse tensors. This makes it a strong alternative to option 1.
5. anything else?

**Comment by cpuhrsch** [comment]
Another option, before we land one some generalized metadata extension that will also work well for all the other operators we intend to support, is to return a COO Tensor. If the indices aren't sorted, the values can definitely be taken as is. The transposition of a CSR matrix would then result in a COO matrix, which isn't ideal, but perhaps less risky and a way to get started.

**Comment by cpuhrsch** [comment]
Summary of offline discussion:

For now we'll invest into [CSR to COO conversion](https://github.com/pytorch/pytorch/issues/57381) and then do transposition via COO until we've settled on the right metadata extension to do CSR transposition natively. While this is a lot more expensive, it doesn't get us stuck in any generalizations we'll later on regret and only works for one operator.

**Comment by albanD** [comment]
> This statement requires clarification: is it a general requirement or is it just a consequence of using strided layout?

@pearu this is a general statement within pytorch.
These operations are documented and handled by the autograd explicitly as sharing data between the input and output. So the user actually expects this behavior. And autograd formulas would be wrong if you don't actually share data.

Note that some backends like xla don't support such memory sharing but still handle view/inplace by making sure that any inplace modification is properly reflected on all other Tensors that are "views" of the one being modified inplace.

**Comment by pearu** [comment]
[Requiring transpose as a view](https://github.com/pytorch/pytorch/issues/61059#issuecomment-877896058) is controversial because CSR transpose permutes the elements of CSR values tensor in a non-trivial way. However, we can write:
```
CSR(crow_indices, col_indices, values).transpose(0, 1) == CSR(new_crow_indices, new_col_indices, matmul(P, values))
```
where `P` is a certain permutation matrix that can be computed together with `new_crow/col_indices` via “CSR indices->COO indices->swap COO indices rows to get transpose indices->coalesce swapped COO indices->new CSR indices”.
Here, `P` can be constructed in the coalesce step (by introducing a third row to indices filled with `range(nnz)` (or equivalently, using auxiliary COO tensor with `range(nnz)` values) that after coalescing (read: lex-sorting indices) contains the permutation indices).
Would this formalism be useful for CSR-matmul autograd? If so, perhaps we should consider making COO transpose as a view of values first because `P` would be an identity matrix.

**Comment by cpuhrsch** [comment]
@pearu  I think we should first introduce coverage to unblock matmul gradients and then fix transposition for sparse matrices together. We can do that right now and it should be a relatively small change. 

In general, what you propose might work. Can you open a separate issue that talks about transposition currently not being a view altogether? At some point I tried to find a concrete example where the current behavior of COO transposition breaks autograd, but COO coverage didn't allow me to find an operator where this would hold. Maybe I didn't look closely enough.

**Comment by pearu** [comment]
@cpuhrsch I agree, implementing CSR transpose as a copy operation would be a step forward.

For the sake of testing autograd with respect to invalidating "transpose is a view operation", I made strided transpose a copy operation (`transpose` implementation returns `result.clone()`). As a result, test_autograd.py has one failure in `test_view_inplace_always_creates_a_view` and 441 successful tests.

@albanD, can you provide a "better" example where autograd breaks when the transpose of a strided tensor is not a view operation ("better" in the sense that the example uses operations that would allow changing the input layout from strided to sparse COO layout and that would lead to the same autograd breakage)?

Here's a issue for discussion non-view transpose vs autograd: TBD.

**Comment by albanD** [comment]
I think it "breaks" in a few ways:
- User expect a view and thus no extra memory use when calling such op
- User expect this op + inplace to modify the original Tensor
- Autograd expect this op + inplace to modify the original Tensor and computes the gradients accordingly. The gradient of the input will be modified by the inplace as-if it was also modified.

**Comment by cpuhrsch** [comment]
We now have [CSC](https://github.com/pytorch/pytorch/pull/77615).
